cli/src/trusted_operation.rs

```diff
@@ -227,6 +223,79 @@ fn send_direct_request(
 	}
 }
 
+/// sends a rpc watch request to the worker api server
+pub fn get_json_request(
+	trusted_args: &TrustedArgs,
+	operation_call: &TrustedOperation,
+	shielding_pubkey: sgx_crypto_helper::rsa3072::Rsa3072PubKey,
+) -> String {
+	let operation_call_encrypted = shielding_pubkey.encrypt(&operation_call.encode()).unwrap();
+	let shard = read_shard(trusted_args).unwrap();
+
+	// compose jsonrpc call
+	let request = Request { shard, cyphertext: operation_call_encrypted };
+	RpcRequest::compose_jsonrpc_call(
+		"author_submitAndWatchExtrinsic".to_string(),
+		vec![request.to_hex()],
+	)
+	.unwrap()
+}
+
+pub fn wait_until(
+	receiver: &Receiver<String>,
+	until: fn(TrustedOperationStatus) -> bool,
+) -> Option<(H256, Instant)> {
+	debug!("waiting for rpc response");
+	loop {
+		match receiver.recv() {
+			Ok(response) => {
+				debug!("received response: {}", response);
+				let parse_result: Result<RpcResponse, _> = serde_json::from_str(&response);
+				if let Ok(response) = parse_result {
+					if let Ok(return_value) = RpcReturnValue::from_hex(&response.result) {
+						debug!("successfully decoded rpc response");
+						match return_value.status {
+							DirectRequestStatus::Error => {
+								debug!("request status is error");
+								if let Ok(value) =
+									String::decode(&mut return_value.value.as_slice())
+								{
+									println!("[Error] {}", value);
+								}
+								return None
+							},
+							DirectRequestStatus::TrustedOperationStatus(status) => {
+								debug!("request status is: {:?}", status);
+								if let Ok(value) = Hash::decode(&mut return_value.value.as_slice())
+								{
+									println!("Trusted call {:?} is {:?}", value, status);
+									if until(status) {
+										return Some((value, Instant::now()))
+									}
+								}
+							},
+							_ => {
+								debug!("request status is ignored");
+								return None
+							},
+						}
```

Comment:
This part is heavily nested, I'm sure it can be improved, e.g. by extracting some functions from it (e.g. the inner part with `match`)

Reply:
Sorry, this cannot be easily extracted. It's also a copy-paste from send_direct_request() and should actually be properly refactored.